atc/exec/check_step.go

```diff
@@ -93,11 +113,91 @@ func (step *CheckStep) run(ctx context.Context, state RunState) error {
 		return fmt.Errorf("resource types creds evaluation: %w", err)
 	}
 
-	timeout, err := time.ParseDuration(step.plan.Timeout)
+	resourceConfig, err := step.resourceConfigFactory.FindOrCreateResourceConfig(step.plan.Type, source, resourceTypes)
+	if err != nil {
+		return fmt.Errorf("create resource config: %w", err)
+	}
+
+	// XXX(check-refactor): we should remove scopes as soon as it's safe to do
+	// so, i.e. global resources is on by default. i think this can be done when
+	// time resource becomes time var source (resolving thundering herd problem)
+	// and IAM is handled via var source prototypes (resolving unintentionally
+	// shared history problem)
+	scope, err := step.delegate.FindOrCreateScope(resourceConfig)
+	if err != nil {
+		return fmt.Errorf("create resource config scope: %w", err)
+	}
+
+	lock, run, err := step.delegate.WaitToRun(ctx, scope)
+	if err != nil {
+		return fmt.Errorf("wait: %w", err)
+	}
+
+	if run {
+		defer func() {
+			err := lock.Release()
+			if err != nil {
+				logger.Error("failed-to-release-lock", err)
+			}
+		}()
+
+		fromVersion := step.plan.FromVersion
+		if fromVersion == nil {
+			latestVersion, found, err := scope.LatestVersion()
+			if err != nil {
+				return fmt.Errorf("get latest version: %w", err)
+			}
+
+			if found {
+				fromVersion = atc.Version(latestVersion.Version())
+			}
+		}
+
+		_, err = scope.UpdateLastCheckStartTime()
+		if err != nil {
+			return fmt.Errorf("update check end time: %w", err)
+		}
+
+		result, err := step.runCheck(ctx, logger, timeout, resourceConfig, source, resourceTypes, fromVersion)
+		if setErr := scope.SetCheckError(err); setErr != nil {
+			logger.Error("failed-to-set-check-error", setErr)
+		}
+		if _, ok := err.(runtime.ErrResourceScriptFailed); ok {
+			step.delegate.Finished(logger, false)
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("run check: %w", err)
+		}
+
+		err = scope.SaveVersions(db.NewSpanContext(ctx), result.Versions)
+		if err != nil {
+			return fmt.Errorf("save versions: %w", err)
+		}
+
+		_, err = scope.UpdateLastCheckEndTime()
+		if err != nil {
+			return fmt.Errorf("update check end time: %w", err)
+		}
+	}
+
+	err = step.delegate.PointToSavedVersions(scope)
```

Comment:
This is the only place we point a resource at its resource_config_scope, correct?

I set a pipeline locally whose resource script fails, and I was expecting to see the check error on the resource page in the UI, but it doesn't come up (it just always says `checked successfully`.

I realized that's because the check error is being set on the scope, but not the resource itself (is the resource check error for failing to start the check?). Since we never point the resource at the scope, though, the check error is "lost".

You mentioned we could stream the output of the build on the resource page - then I guess we wouldn't need `resource_config_scopes.check_error`, right?

Reply:
Ah, yeah good catch. I wonder if we're missing higher-level test coverage for this. Seems like it would prevent seeing check errors in general.

Some context: the scope is set after versions are found to make sure there isn't a window of time where a resource is pointing to an empty scope after e.g. a credential rotation or new resource type version.

I can change it to point to the new scope even if checking fails, the code might just look a little crusty.

In the long run, yeah, I think just showing the output/build status on the resource page would be a lot cleaner, since then we can get rid of this whole `check_error` thing. We could also return a resource's current build in the API to make the resource orange in the pipeline/dashboard UI. (Though maybe its color could instead reflect the build status for consistency's sake? 🤔)

This is definitely a blocker since it'll probably mean users don't see check failures at all in the UI. I think I'll push a quick fix first (set scope on check failure) to get the PR unblocked and then spike on the build UI stuff on another branch.